launcher/InstanceImportTask.cpp

```diff
@@ -108,76 +108,77 @@ void InstanceImportTask::downloadFromUrl()
     filesNetJob->start();
 }
 
-QString InstanceImportTask::getRootFromZip(QStringList files)
-{
-    if (!isRunning()) {
-        return {};
-    }
-    auto cleanPath = [](QString path) {
-        if (path == ".")
-            return QString();
-        QString result = path;
-        if (result.startsWith("./"))
-            result = result.mid(2);
-        return result;
-    };
-    for (auto&& fileName : files) {
-        setDetails(fileName);
-        QFileInfo fileInfo(fileName);
-        if (fileInfo.fileName() == "instance.cfg") {
-            qDebug() << "MultiMC:" << true;
-            m_modpackType = ModpackType::MultiMC;
-            return cleanPath(fileInfo.path());
-        }
-        if (fileInfo.fileName() == "manifest.json") {
-            qDebug() << "Flame:" << true;
-            m_modpackType = ModpackType::Flame;
-            return cleanPath(fileInfo.path());
-        }
-        QCoreApplication::processEvents();
-    }
-    return {};
-}
-
 void InstanceImportTask::processZipPack()
 {
     setStatus(tr("Attempting to determine instance type"));
+    setDetails("");
     QDir extractDir(m_stagingPath);
     qDebug() << "Attempting to create instance from" << m_archivePath;
 
     // open the zip and find relevant files in it
     MMCZip::ArchiveReader packZip(m_archivePath);
-    if (!packZip.collectFiles()) {
-        emitFailed(tr("Unable to open supplied modpack zip file."));
-        return;
-    }
-
     qDebug() << "Attempting to determine instance type";
 
-    QString root;
+    auto cleanPath = [](QString path) {
+        if (path == ".")
+            return QString();
+        QString result = path;
+        if (result.startsWith("./"))
+            result = result.mid(2);
+        return result;
+    };
 
+    QString root;
     // NOTE: Prioritize modpack platforms that aren't searched for recursively.
     // Especially Flame has a very common filename for its manifest, which may appear inside overrides for example
     // https://docs.modrinth.com/docs/modpacks/format_definition/#storage
-    if (packZip.exists("/modrinth.index.json")) {
-        // process as Modrinth pack
-        qDebug() << "Modrinth:" << true;
-        m_modpackType = ModpackType::Modrinth;
-    } else if (packZip.exists("/bin/modpack.jar") || packZip.exists("/bin/version.json")) {
-        // process as Technic pack
-        qDebug() << "Technic:" << true;
-        extractDir.mkpath("minecraft");
-        extractDir.cd("minecraft");
-        m_modpackType = ModpackType::Technic;
-    } else {
-        root = getRootFromZip(packZip.getFiles());
-        setDetails("");
+    if (!packZip.parse([this, &extractDir, &root, cleanPath](MMCZip::ArchiveReader::File* f, bool& stop) {
```

Comment:
nitpick: maybe don't inline a whole lambda here and stuff it in an auto var instead or keep it as a separate func

Reply:
done